framework/src/node/consensus/certificate_generation/commit_pool.ts

```diff
@@ -62,15 +64,103 @@ export class CommitPool {
 
 	// eslint-disable-next-line @typescript-eslint/no-empty-function
 	public async job(): Promise<void> {}
-	// eslint-disable-next-line @typescript-eslint/no-empty-function
-	public addCommit(_commit: SingleCommit, _height: number): void {}
-	// eslint-disable-next-line @typescript-eslint/no-empty-function
-	public validateCommit(_commit: SingleCommit): boolean {
+
+	public addCommit(commit: SingleCommit, height: number): void {
+		const currentCommits = this._nonGossipedCommits.get(height) ?? [];
+		const doesCommitExist = currentCommits.some(aCommit =>
+			aCommit.certificateSignature.equals(commit.certificateSignature),
+		);
+		if (!doesCommitExist) {
+			this._nonGossipedCommits.set(height, [...currentCommits, commit]);
+		}
+	}
+
+	public async validateCommit(apiContext: APIContext, commit: SingleCommit): Promise<boolean> {
+		// Validation step 4
+		const blockHeaderAtCommitHeight = await this._chain.dataAccess.getBlockHeaderByHeight(
+			commit.height,
+		);
+		if (!blockHeaderAtCommitHeight.id.equals(commit.blockID)) {
+			return false;
+		}
+
+		// Validation step 1
+		const doesCommitExistsInNonGossipedCommits = !!this._nonGossipedCommits
+			.get(commit.height)
+			?.some(
+				nonGossipedCommit =>
+					nonGossipedCommit.blockID === commit.blockID &&
+					nonGossipedCommit.validatorAddress === commit.validatorAddress,
+			);
+
+		const doesCommitExistsInGossipedCommits = !!this._gossipedCommits
+			.get(commit.height)
+			?.some(
+				gossipedCommit =>
+					gossipedCommit.blockID === commit.blockID &&
+					gossipedCommit.validatorAddress === commit.validatorAddress,
+			);
+
+		const doesCommitExist =
+			doesCommitExistsInGossipedCommits || doesCommitExistsInNonGossipedCommits;
+
+		if (doesCommitExist) {
+			return false;
+		}
+
+		// Validation Step 2
+		const maxRemovalHeight = await this._getMaxRemovalHeight();
+		if (commit.height <= maxRemovalHeight) {
+			return false;
+		}
+
+		// Validation Step 3
+		const { maxHeightPrecommitted } = await this._bftAPI.getBFTHeights(apiContext);
+		const isCommitInRange =
+			commit.height >= maxHeightPrecommitted - COMMIT_RANGE_STORED &&
+			commit.height < maxHeightPrecommitted;
+		const doesBFTParamExistForNextHeight = await this._bftAPI.existBFTParameters(
+			apiContext,
+			commit.height + 1,
+		);
+		if (!isCommitInRange && !doesBFTParamExistForNextHeight) {
+			return false;
+		}
+
+		// Validation Step 5
```

Comment:
Validation step 4 is missing

Reply:
it's at the beginning  of the validaCommit method 😄 (As per Jan's suggestion). May you check that the combination of validation step 4 and validation step 2 in the code is valid for validation step 2 in the LIP? I am not 100% certain about it.

Reply:
ah missed out the validation step in the beginning, I think step 2 and step 4 can be doesn't need a combination so its fine